Feature: Candle conversion & Candle validation

gloriousCode opened the pull request:
![image](https://thumbs.gfycat.com/HoarseConstantHorseshoebat-size_restricted.gif)
_select * from candle_

# PR Description

This is an expansion to the data history manager so that we can convert and validate data at scale. This _may_ give a quant confidence in the data they are using to build strategies for the backtester. Here are the new features
+ Convert trades to candles job type
+ Convert candles to new candles job type
+ Validate candles against exchange API data job type
+ Validate candles against a secondary exchange's API data job type
+ Job queuing for large data set acquisition/conversion/validation
+ Pausing and unpausing jobs
+ Candle data can be associated with its source job
+ Candle data can be associated with its validation job
  + Candle data will contain any issues that were found when validating
+ Backtester will warn if any candle data it uses has a validation issue
+ Fix an issue with Binance trade fetching where we would miss many initial trades (which accounts for most of the line changes in the PR)

#### Testing
Please ensure you have run dbmigrate before testing

### Documentation
The following has been taken from the readme to highlight changes. Please [read it here](https://github.com/gloriousCode/gocryptotrader/blob/candlekeep-mysteries/engine/datahistory_manager.md) for full details

### Job types:
| Command | Description | DataHistoryJobDataType |
| ------- | ----------- | ---------------------- |
| savecandles | Will fetch candle data from an exchange and save it to the database | 0 |
| savetrades | Will fetch trade data from an exchange and save it to the database | 1 |
| converttrades | Convert trades saved to the database to any candle resolution eg 30min | 2 |
| convertcandles | Convert candles saved to the database to a new resolution eg 1min -> 5min | 3 |
| validatecandles | Will compare database candle data with API candle data - useful for validating converted trades and candles | 4 |
| secondaryvalidatecandles | Will compare database candle data with a different exchange's API candle data | 5 |


### Job status types

| Job Status | Description | Representative value |
| ---------- | ----------- | -------------------- |
| active | A job that is ready to processed | 0 |
| failed | The job has failed to retrieve/covert/validate the data you have specified. See the associated data history job results to understand why | 1 |
| complete | The job has successfully retrieved/converted/validated all data you have specified | 2 |
| removed | The job has been deleted. No data is removed, but the job can no longer be processed | 3 |
| missing data | The job is complete, however there is some missing data. See the associated data history job results to understand why | 4 |
| paused | The job has been paused and will not be processed. Either it has a prerequisite job that needs to be completed, or a user must unpause the job | 5 |

### Job queuing and prerequisite jobs
You can add jobs which will be paused by default by using the `prerequisite` subcommand containing the associated job nickname. The prerequisite job will be checked to ensure it exists and has not yet completed and add the relationship.
+ Once you have set a prerequisite job, when the prerequisite job status is set to `complete`, the data history manager will search for any jobs which are pending its completion and update their status to `active`.
+ If the prerequisite job is deleted or fails, the upcoming job will _not_ be run.
+ Multiple jobs can use the same prerequisite job, but a job cannot have multiple prerequisites
  + Attempting to add a new prerequisite will overwrite the existing prerequisite
+ You can chain many queued jobs together allowing for automated and large scale data retrieval projects. For example:

| Job Type | Job Name | Description | Prerequisite Job |
| -------- | -------- | ----------- | ---------------- |
| savetrades | save-trades | Save jobs between 01-01-2021 and 01-02-2021 | |
| converttrades | convert-trades | Convert trades to 5m candles | save-trades |
| validatecandles | validate-candles | Ensure the converted trades match the exchange's API data | convert-trades |
| convertcandles | convert-candles-10m | Now that we have confidence in conversion, convert candle to 10m | validate-candles |
| convertcandles | convert-candles-25m | Now that we have confidence in conversion, convert candle to 25m | validate-candles |
| convertcandles | convert-candles-1d | Now that we have confidence in conversion, convert candle to 1d | validate-candles |

### Backtester warnings
Any candles used that have had validation issues will be displayed in the warning table to allow an esteemed quant to make an informed decision on whether their results can truly be trustworthy:
![image](https://user-images.githubusercontent.com/9261323/126089527-4a464359-8d26-4ffc-ac9f-62a77b3d7616.png)




- [x] New feature (non-breaking change which adds functionality)

## How has this been tested
Coverage on all new functionality
- [x] go test ./... -race
- [x] golangci-lint run

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [x] Any dependent changes have been merged and published in downstream modules

shazbert commented on engine/datahistory_manager.go:
```diff
@@ -438,29 +477,531 @@ completionCheck:
+			newJobNames = append(newJobNames, newJobs[i].Nickname)
+		}
+		if len(newJobNames) > 0 {
+			log.Infof(log.DataHistory, "setting the follow jobs to active: %s", strings.Join(newJobNames, ", "))
```
follow -> following

shazbert commented on engine/datahistory_manager.go:
```diff
@@ -598,15 +1194,39 @@ func (m *DataHistoryManager) validateJob(job *DataHistoryJob) error {
-		job.Interval = defaultDataHistoryTradeInterval
+	if job.DataType == dataHistoryTradeDataType {
+		if job.Interval >= kline.FourHour || job.Interval <= kline.OneMin {
+			log.Warnf(log.DataHistory, "job %s interval %v outside limit of 4h, defaulting to %v", job.Nickname, job.Interval.Word(), defaultDataHistoryTradeInterval)
```
add 1m to string

shazbert commented on engine/datahistory_manager.go:
```diff
@@ -598,15 +1194,39 @@ func (m *DataHistoryManager) validateJob(job *DataHistoryJob) error {
-		log.Warnf(log.DataHistory, "job %s interval %v outside limits, defaulting to %v", job.Nickname, job.Interval.Word(), defaultDataHistoryTradeInterval)
-		job.Interval = defaultDataHistoryTradeInterval
+	if job.DataType == dataHistoryTradeDataType {
+		if job.Interval >= kline.FourHour || job.Interval <= kline.OneMin {
```
shouldn't 4hr and one min be acceptable?

shazbert commented on engine/datahistory_manager.go:
```diff
@@ -598,15 +1194,39 @@ func (m *DataHistoryManager) validateJob(job *DataHistoryJob) error {
+			job.Interval = defaultDataHistoryTradeInterval
+		}
+		if job.RequestSizeLimit > 20 {
+			log.Warnf(log.DataHistory, "job %s request size %v outside limit of 100, defaulting to %v", job.Nickname, job.RequestSizeLimit, defaultDataHistoryTradeRequestSize)
```
100 -> 20?

shazbert commented on engine/datahistory_manager.go:
```diff
@@ -598,15 +1194,39 @@ func (m *DataHistoryManager) validateJob(job *DataHistoryJob) error {
+			job.DecimalPlaceComparison = defaultDecimalPlaceComparison
+		}
+		if job.RequestSizeLimit > 500 {
+			log.Warnf(log.DataHistory, "job %s validation batch limit %v too high. defaulting to %v", job.Nickname, job.RequestSizeLimit, defaultDataHistoryRequestSizeLimit)
```
higher than 500

shazbert commented on engine/rpcserver.go:
```diff
@@ -3339,18 +3339,26 @@ func (s *RPCServer) UpsertDataHistoryJob(_ context.Context, r *gctrpc.UpsertData
+		DataType:                 dataHistoryDataType(r.DataType),
+		MaxRetryAttempts:         r.MaxRetryAttempts,
+		Status:                   dataHistoryStatusActive,
+		CreatedDate:              time.Time{},
```
rm?

shazbert commented on exchanges/exchange.go:
```diff
@@ -433,6 +433,7 @@ func (b *Base) GetRequestFormattedPairAndAssetType(p string) (currency.Pair, ass
 			}
 		}
 	}
+
```
😡

shazbert commented on exchanges/kline/kline.go:
```diff
@@ -331,6 +331,68 @@ func (i *Interval) IntervalsPerYear() float64 {
+	var candles []Candle
+	for i := range item.Candles {
+		candles = append(candles, item.Candles[i])
+		thing := int64(i + 1)
```
![image](https://user-images.githubusercontent.com/9391715/126249805-bcc365a1-a08c-4cb5-9545-8c40a5bbcc1f.png)

shazbert reviewed (changes requested):
First pass, code looks really solid! 🦅 Just some things I picked up along the way before I deep dive into testing. There was some missed test code from code cov coming up that may need to be addressed.

gloriousCode commented on database/migrations/20210630125121_datahistoryjob_conversions/postgres.sql:
```diff
@@ -0,0 +1,19 @@
+
+-- +goose Down
+ALTER TABLE datahistoryjob
+    DROP conversion_interval,
+    DROP overwrite_data,
+    DROP decimal_place_comparison,
+    DROP CONSTRAINT datahistoryjob_secondary_exchange_id_fkey,
+    DROP secondary_exchange_id ,
+    DROP issue_tolerance_percentage,
+    DROP replace_on_issue;
```
What do you mean?

shazbert replied:
Look maybe its not really important but I like to do previous command first and iterate through backwards from whats been done. Just resolve if you think this is unnecessary.

gloriousCode replied:
Thanks for the explanation! Makes sense

shazbert commented on cmd/gctcli/data_history.go:
```diff
@@ -139,34 +63,228 @@ var dataHistoryCommands = &cli.Command{
+			Action: upsertDataHistoryJob,
+		},
+		{
+			Name:   "savetrades",
```
```json
--request_size_limit value  the number of candles to retrieve per API request (default: 500)
```

Might be better to change to trades?

Also im confused by this debug output. Does this mean only 5 trades are loaded?
```json
job ftxtest4 request size 500 outside limit of 5, defaulting to 5
```

gloriousCode replied:
Its a bit more complicated, but highlighted here: https://github.com/thrasher-corp/gocryptotrader/blob/master/engine/datahistory_manager.md#candle-intervals-and-trade-fetching 
5 intervals worth of trades will be fetched. But I'll rethink things

shazbert replied:
Thanks for clarifying!

shazbert commented on cmd/gctcli/data_history.go:
```diff
@@ -139,34 +63,228 @@ var dataHistoryCommands = &cli.Command{
+			Action: upsertDataHistoryJob,
+		},
+		{
+			Name:   "converttrades",
```
Wondering if these flags are needed in this context?
```json
--request_size_limit value  the number of candles to retrieve per API request (default: 500)
   --max_retry_attempts value  the maximum retry attempts for an interval period before giving up (default: 3)
   --batch_size value          the amount of API calls to make per run (default: 3)
   --overwrite_existing_data   will process and overwrite data if matching data exists at an interval period. if false, will not process or save data (default: false)
```

gloriousCode replied:
Yes they do. I can update the descriptions to be more relevant, but they all use the same parameters.
- request size limit helps calculate the data ranges to fetch
- max retry attempts is still valid as when you create a job, there's no guarantee the data is already there, and a job may be running at the same time to save said data its trying to convert - that onus is on the user.
- batch size dictates how many cycles per processing of a job to run. We don't convert all trades in one go to be considerate of resources
- overwrite existing data is important as there may already be saved data that differs at the destination

shazbert replied:
Okay makes sense now! 🚀

shazbert reviewed (changes requested):
Thanks for making those changes before, just a few more potential issues I have spotted. 👍

gloriousCode commented on engine/datahistory_manager_test.go:
```diff
@@ -964,6 +983,7 @@ func createDHM(t *testing.T) *DataHistoryManager {
 		tradeChecker:    dataHistoryHasDataChecker,
 		candleLoader:    dataHistoryCandleLoader,
 		interval:        time.NewTicker(time.Minute),
+		verbose:         true,
```
codecov complaining about lack of coverage of verbose logs 😄 has no impact, happy to revert

shazbert replied:
Yeah, never mind I thought it might be superfluous.

shazbert reviewed (changes requested):
```json
[DEBUG] | DATAHISTORY | 21/07/2021 14:32:13 | Data history manager shutdown.
[ERROR] | DATABASE | 21/07/2021 14:32:31 | Database connection error:sql: database is closed
[DEBUG] | DATABASE | 21/07/2021 14:32:31 | Database manager shutdown.
[DEBUG] | DISPATCH | 21/07/2021 14:32:31 | Dispatch manager shutting down...
[DEBUG] | DISPATCH | 21/07/2021 14:32:31 | Dispatch manager shutdown.
[ERROR] | DATAHISTORY | 21/07/2021 14:32:31 | ftxtest5 could not load candle data: sqlite3: failed to assign all query results to Candle slice: bind failed to execute query: sql: database is closed
```
Doesn't really like shutting down when it's processing.

gloriousCode commented:
> Doesn't really like shutting down when it's processing.

So this is a result of using a go routine to handle job processing. On the upside, when you shutdown, there is no waiting for a job to finish processing and the application will just close. Additionally, any job not completed will simply get run the next time, so no real harm to it. The consequence is that the go routine is still running and will attempt to utilise other subsystems that are shut down and you get sweet errors like that.

I've add a few more `IsRunning()` checks to more functions, this should minimise the chance of such an error popping up. Happy to continue discussion on it though

shazbert reviewed (approved):
tested ACK. Everything looks good, tested the prerequisites. Shutting down does not generate any errors! 🚀

thrasher- commented on cmd/documentation/engine_templates/datahistory_manager.tmpl:
```diff
@@ -20,9 +32,28 @@
+### How do I add a job?
++ First ensure that the data history monitor is enabled, you can do this via the config (see table `dataHistoryManager` under Config parameters below), via run time parameter (see table Application run time parameters below) or via the RPC command `enablesubsystem --subsystemname="data_history_manager"`
++ The simplest way of adding a new data history job is via the GCTCLI under `/cmd/gctcli`.
+  + Modify the following example command to your needs: `.\gctcli.exe datahistory addjob savecandles --nickname=binance-spot-bnb-btc-1h-candles --exchange=binance --asset=spot --pair=BNB-BTC --interval=3600 --start_date="2020-06-02 12:00:00" --end_date="2020-12-02 12:00:00" --request_size_limit=10 --data_type=0 --max_retry_attempts=3 --batch_size=3`
```
--data_type no longer exists

thrasher- reviewed (changes requested):
Had a lovely an indefinite lock when attempting the following:

```
./gctcli datahistory addjob savetrades --nickname=binance-spot-btc-usdt-2021-trades --exchange=binance --asset=spot --pair=btc-usdt --start_date="2021-07-01 00:00:00" --end_date="2021-07-04 00:00:00" --interval=86400

./gctcli datahistory getactivejobs
outcome: worked initially

./gctcli datahistory deletejob 0e4c3b07-f05a-40d2-bf4f-8017af8eb830
{
 "status": "success"
}
```

Attempting to add another job or using getactivehistory will hang the requester client indefinitely 😄

thrasher- commented:
> ```json
> [DEBUG] | DATAHISTORY | 21/07/2021 14:32:13 | Data history manager shutdown.
> [ERROR] | DATABASE | 21/07/2021 14:32:31 | Database connection error:sql: database is closed
> [DEBUG] | DATABASE | 21/07/2021 14:32:31 | Database manager shutdown.
> [DEBUG] | DISPATCH | 21/07/2021 14:32:31 | Dispatch manager shutting down...
> [DEBUG] | DISPATCH | 21/07/2021 14:32:31 | Dispatch manager shutdown.
> [ERROR] | DATAHISTORY | 21/07/2021 14:32:31 | ftxtest5 could not load candle data: sqlite3: failed to assign all query results to Candle slice: bind failed to execute query: sql: database is closed
> ```
> 
> Doesn't really like shutting down when it's processing.

Looks like I got the same after a attempting a Ctrl + C with an active job (same job previously mentioned). It took about 5 minutes or so to get the shutdown confirmation (using sqlite also):

```
INFO]  | 02/08/2021 16:16:18 | processing data history jobs
[ERROR] | 02/08/2021 16:16:18 | subsystem not started
[INFO]  | 02/08/2021 16:16:18 | completed run of data history jobs
[ERROR] | 02/08/2021 16:16:18 | Database connection error:sql: database is closed
[DEBUG] | 02/08/2021 16:16:18 | Database manager shutdown.
[ERROR] | 02/08/2021 16:16:18 | sqlite3: failed to execute a one query for datahistoryjob: bind failed to execute query: sql: database is closed
```

gloriousCode commented:
The hanging issue you were experiencing was a result of inefficiencies dealing with large amounts of data. The function to verify whether trade data existed slowed to a crawl once > 1m trades in the database and blocked shutdown + SQL access if SQLite. Similarly, when saving all trades to the database in one transaction, it would block until complete. 

I've changed two things:
- Added a new config field `maxResultInsertions` to save data in specified chunks, with the default being 10,000 results and will loop until complete. This is for trade and candle data via the data history manager.
- Changed the way trade data is verified. It now uses the datahistory job results. If a result range is marked as complete, then the job will process the next incomplete range.

shazbert reviewed (approved):
Changes ACK - 
```json
[DEBUG] | DATAHISTORY | 04/08/2021 14:49:41 | running data history job ftx-spot-btc-usdt-2021-trades start: 2021-06-30 19:00:00 +0000 UTC end: 2021-06-30 21:30:00 +0000 UTC interval: 15m0s datatype: trades
[DEBUG] | DATAHISTORY | 04/08/2021 14:49:41 | job ftx-spot-btc-usdt-2021-trades processing range {2021-06-30 19:00:00 +0000 UTC 1625079600}-{2021-06-30 21:30:00 +0000 UTC 1625088600}
[DEBUG] | DATAHISTORY | 04/08/2021 14:49:47 | Saving 2336 trades. Range 0-2336/2336
```

Populating the DB well! No signs of slowdown and closes fine when ctrl-c. 🚀

thrasher- reviewed (approved):
Issues now resolved, tACK. Great stuff @gloriousCode !
